Adding support for NCrystal materials in OpenMC

marquezj opened the pull request:
After some discussion with @paulromano we cleaned up this version of OpenMC with support for [NCrystal](https://github.com/mctools/ncrystal). NCrystal is a software library that provides low energy neutron physics to Monte Carlo codes. It was initially designed with Geant4 in mind, but it currently has interfaces with many other codes, and it can be called from Python, C and C++. NCrystal is developed mainly at the European Spallation Source. You can read more about NCrystal in these papers:

  X.-X. Cai and T. Kittelmann, NCrystal: A library for thermal neutron
  transport, Computer Physics Communications 246 (2020) 106851,
  https://doi.org/10.1016/j.cpc.2019.07.015

  T. Kittelmann and X.-X. Cai, Elastic neutron scattering models
  for NCrystal, Computer Physics Communications 267 (2021) 108082,
  https://doi.org/10.1016/j.cpc.2021.108082

  X.-X. Cai, T. Kittelmann, et. al., "Rejection-based sampling of inelastic
  neutron scattering", Journal of Computational Physics 380 (2019) 400-407,
  https://doi.org/10.1016/j.jcp.2018.11.043

or watch the presentation from @tkittel, at ND2022:

https://indico.frib.msu.edu/event/52/contributions/954/

We have used NCrystal together with NJOY to generate a large library of new materials in ACE format (the [NJOY-NCrystal library](https://github.com/highness-eu/NJOY-NCrystal-Library/)), but calling NCrystal directly allows to introduce in OpenMC additional physics that is not supported by the ACE format.

In particular, NCrystal allows to generate and sample on the fly scattering kernels for solid polycrystalline materials at any temperature. This is done at initialization, based on a human readable description stored in text form in the [NCMAT](https://github.com/mctools/ncrystal/wiki/NCMAT-format) format. For existing ENDF-6 scattering kernels, it allows the direct sampling of S(alpha, beta), which allows for continuous angular distributions (which are still not supported by the ACE format). NCrystal also support oriented materials (single crystals, highly-oriented pyrolytic graphite) and has plug-in support for other physics, currently being used to include small angle neutron scattering and inelastic magnetic scattering. All this would make OpenMC more useful to the neutron scattering community, while adding advanced nuclear data capabilities for the reactor community.

This pull request is planned as an optional feature that could be activated with cmake, using the [OPENMC_USE_NCRYSTAL](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/CMakeLists.txt#L39) flag after [installing and initializing](https://github.com/mctools/ncrystal/blob/master/INSTALL) NCrystal. A new method is added to the Material class in Python, [to create materials from NCrystal](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/openmc/material.py#L317), which takes the composition and density, creates the OpenMC material and associates it with NCrystal.

After reading the xml files, [an NCrystal object](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/src/material.cpp#L67) is created in memory, which provides [thermal scattering cross sections](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/src/material.cpp#L806) and [scattering events](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/src/physics.cpp#L150) (for the calculation of the scattering events, the direction of the neutron is a parameter to be able to consider oriented materials). When the thermal scattering cross section is computed, the [nuclear elastic scattering component is removed](https://github.com/highness-eu/openmc/blob/55e8bd45c24e876a8df02a19fcca56158f091bf3/src/material.cpp#L806), similarly to what is done in the standard S(alpha, beta) treatment.

There a couple of open topics that we could not decide, and are marked with "fixme" in the code. For instance, NCrystal uses a cache that could be implemented in OpenMC to reduce the computation time. Also, new tags for NCrystal events could be added for tallies. But, overall, we think it is ready for merging upstream.

tkittel commented:
I can add to what @marquezj wrote, that since last week NCrystal is also available on conda-forge (since @paulromano suggested this during our meeting).

paulromano commented:
Thanks for the PR @marquezj! Really excited to see this. Just wanted to give you a heads up I probably won't get to reviewing this until early next week.

tkittel commented on CMakeLists.txt:
```diff
@@ -98,6 +99,14 @@ macro(find_package_write_status pkg)
+#===============================================================================
+
+if(OPENMC_USE_NCRYSTAL)
+  find_package(NCrystal REQUIRED PATH_SUFFIXES ./)#fixme
```
Concerning this fixme: We were not sure about the `PATH_SUFFIXES ./` part. We saw `PATH_SUFFIXES` being used elsewhere in the file, but normally we would expect a `find_package(NCrystal)` to be sufficient.

tkittel commented on src/physics.cpp:
```diff
@@ -140,8 +140,26 @@ void sample_neutron_reaction(Particle& p)
+  if (model::materials[p.material()]->m_NCrystal_mat() != nullptr && p.E() < settings::ncrystal_max_energy){
+    NcrystalRNG_Wrapper rng(p.current_seed()); // Initialize RNG
+    //create a cache pointer for multi thread physics
+    NCrystal::CachePtr dummyCache;//fixme: avoid recreating here (triggers malloc)
```
Concerning this fixme: The code here works, but will incur a CPU overhead which is not strictly needed. Ideally each material and each thread would have its own `NCrystal::CachePtr` object, that would be kept around and passed to the `sampleScatter(..)` call in this function. If we could be 100% sure to not have to care about multi-thread safety, we could simply add an `NCrystal::CachePtr` object on the openmc `Material` object next to the shared pointer for the NCrystal process that we have there now (i.e. inside the `ifdef NCRYSTAL` section of material.h).

We left the code as is for now, since it is certainly safe and will give correct results (it will merely make NCrystal look slower than it actually is) and it seemed more beneficial to try to move ahead with the NCrystal+OpenMC integration since the issue is almost certainly ortogonal to any other potential issues concering this integration.

When openmc experts have time to discuss with us how they would like per-material per-thread caches kept (thread-local storage on `Material`?), we can create a new PR addressing this.

paulromano replied:
@tkittel I think this is the right approach for now. Let's stick with the slow but safe approach -- once this is merged, we can create an issue to discuss possible implementations of a per-material, per-thread cache.

tkittel commented:
@paulromano I added comments concerning the two non-obvious fixmes in the PR. Let us know if you have any specific wishes concerning this (or anything else), and we will try to update the PR.

paulromano commented on include/openmc/physics.h:
```diff
@@ -101,6 +105,29 @@ void sample_secondary_photons(Particle& p, int i_nuclide);
+    uint64_t* openmc_seed_;
+public:
+    constexpr NcrystalRNG_Wrapper(uint64_t* s) noexcept : openmc_seed_(s) {}
+  //Can be cheaply created on the stack just before being used in calls to
+  //ProcImpl::Scatter objects, like:
+  //
+  //  RNG_Wrapper rng(seed);
+  //
```
I would remove this / mention it in the comment above the class altogether

paulromano commented on include/openmc/physics.h:
```diff
@@ -101,6 +105,29 @@ void sample_secondary_photons(Particle& p, int i_nuclide);
+class NcrystalRNG_Wrapper : public NCrystal::RNGStream {
+    uint64_t* openmc_seed_;
+public:
+    constexpr NcrystalRNG_Wrapper(uint64_t* s) noexcept : openmc_seed_(s) {}
```
```suggestion
    constexpr NcrystalRNG_Wrapper(uint64_t* seed) noexcept : openmc_seed_(seed) {}
```

paulromano commented on include/openmc/physics.h:
```diff
@@ -101,6 +105,29 @@ void sample_secondary_photons(Particle& p, int i_nuclide);
+  //  RNG_Wrapper rng(seed);
+  //
+protected:
+  //double actualGenerate() override { return prn(openmc_seed_); }
```
```suggestion
```

paulromano commented on include/openmc/material.h:
```diff
@@ -151,12 +155,22 @@ class Material {
   vector<int> nuclide_;                 //!< Indices in nuclides vector
   vector<int> element_;                 //!< Indices in elements vector
+#ifdef NCRYSTAL
+  std::string cfg_; //!< NCrystal configuration string
+  std::shared_ptr<const NCrystal::ProcImpl::Process> m_NCrystal_mat_;
```
See [here](https://docs.openmc.org/en/latest/devguide/styleguide.html#naming) for rules on naming. For member variables, we always use lower-case with underscores. I would also add a "ncrystal" prefix for the `cfg_` variable:
```suggestion
  std::string ncrystal_cfg_; //!< NCrystal configuration string
  std::shared_ptr<const NCrystal::ProcImpl::Process> ncrystal_mat_;
```

paulromano commented on openmc/material.py:
```diff
@@ -109,6 +109,7 @@ def __init__(self, material_id=None, name='', temperature=None):
         self._volume = None
         self._atoms = {}
         self._isotropic = []
+        self._NCrystal_cfg = None
```
```suggestion
        self._ncrystal_cfg = None
```

paulromano commented on include/openmc/material.h:
```diff
@@ -151,12 +155,22 @@ class Material {
+#ifdef NCRYSTAL
+  //! Gwet pointer to NCrystal material object
+  //! \return Pointer to NCrystal material object
+  std::shared_ptr<const NCrystal::ProcImpl::Process> m_NCrystal_mat() const {return m_NCrystal_mat_; };
```
```suggestion
  std::shared_ptr<const NCrystal::ProcImpl::Process> ncrystal_mat() const {return ncrystal_mat_; };
```

paulromano commented on openmc/material.py:
```diff
@@ -131,6 +132,8 @@ def __repr__(self):
 
         string += '{: <16}\n'.format('\tS(a,b) Tables')
 
+        string += '{: <16}=\t{}\n'.format('\tNCrystal conf', self._NCrystal_cfg)
```
```suggestion
        string += '{: <16}=\t{}\n'.format('\tNCrystal conf', self._ncrystal_cfg)
```

paulromano commented on openmc/material.py:
```diff
@@ -310,6 +313,47 @@ def from_hdf5(cls, group: h5py.Group):
         return material
 
+    @classmethod
+    def from_NCrystal(cls, cfg):
```
```suggestion
    def from_ncrystal(cls, cfg):
```

paulromano commented on openmc/material.py:
```diff
@@ -310,6 +313,47 @@ def from_hdf5(cls, group: h5py.Group):
+
+        """
+
+        try:
+            import NCrystal
+        except ImportError:
+            raise SystemExit("ERROR: NCrystal Python module is required.")
```
We normally just allow the default exception thrown by Python
```suggestion
        import NCrystal
```

paulromano commented on openmc/material.py:
```diff
@@ -310,6 +313,47 @@ def from_hdf5(cls, group: h5py.Group):
+        except ImportError:
+            raise SystemExit("ERROR: NCrystal Python module is required.")
+
+        NC_mat = NCrystal.createInfo(cfg)
+        NC_comp = NC_mat.getComposition()
```
On the Python side, we also use lower-case with underscores for local variables
```suggestion
        nc_mat = NCrystal.createInfo(cfg)
        nc_comp = NC_mat.getComposition()
```

paulromano commented on openmc/material.py:
```diff
@@ -310,6 +313,47 @@ def from_hdf5(cls, group: h5py.Group):
+                raise ValueError('NCrystal-OpenMC interface only works with natural elements for now.')
+            material.add_element(atom.elementName(), frac, 'ao')
+
+        material._NCrystal_cfg = cfg
+        material._density_units = "g/cm3"
+        material._density = NC_mat.getDensity()
+        material.temperature = NC_mat.getTemperature()
```
```suggestion
        material._ncrystal_cfg = cfg
        material.set_density("g/cm3", nc_mat.getDensity())
        material.temperature = nc_mat.getTemperature()
```

paulromano commented on openmc/material.py:
```diff
@@ -588,6 +635,9 @@ def add_element(self, element: str, percent: float, percent_type: str = 'ao',
             raise ValueError("Element name should be given by the "
                              "element's symbol or name, e.g., 'Zr', 'zirconium'")
 
+        if self._NCrystal_cfg is not None:
+            raise ValueError("Cannot add elements to NCrystal material")
```
```suggestion
        if self._ncrystal_cfg is not None:
            raise ValueError("Cannot add elements to a NCrystal material")
```

paulromano commented on openmc/material.py:
```diff
@@ -1114,6 +1164,14 @@ def to_xml_element(self):
         if self._volume:
             element.set("volume", str(self._volume))
 
+        if self._NCrystal_cfg:
```
```suggestion
        if self._ncrystal_cfg:
```

paulromano commented on openmc/material.py:
```diff
@@ -1114,6 +1164,14 @@ def to_xml_element(self):
+            if self._macroscopic is not None:
+                raise ValueError("NCrystal materials are not compatible macroscopic cross sections.")
+
+            element.set("cfg", str(self._NCrystal_cfg))
```
```suggestion
            element.set("cfg", str(self._ncrystal_cfg))
```

paulromano commented on src/material.cpp:
```diff
@@ -60,6 +60,14 @@ Material::Material(pugi::xml_node node)
+#ifdef NCRYSTAL
+  if (check_for_node(node, "cfg")) {
+    cfg_ = get_node_value(node, "cfg");
+    write_message(5, "NCrystal config string: >>{}<< ", cfg_);
+    m_NCrystal_mat_  = NCrystal::FactImpl::createScatter(cfg_);
```
```suggestion
    write_message(5, "NCrystal config string: '{}'", ncrystal_cfg_);
    ncrystal_mat_  = NCrystal::FactImpl::createScatter(ncrystal_cfg_);
```

paulromano commented on src/material.cpp:
```diff
@@ -792,6 +800,17 @@ void Material::calculate_neutron_xs(Particle& p) const
+
+  if (m_NCrystal_mat_ != nullptr && p.E() < settings::ncrystal_max_energy){
+    // Calculate scattering XS per atom with NCrystal, only once per material
+    NCrystal::CachePtr dummyCache;
+    auto nc_energy = NCrystal::NeutronEnergy{p.E()};
+    ncrystal_xs = m_NCrystal_mat_->crossSection( dummyCache, nc_energy,  {p.u().x, p.u().y, p.u().z}).get();
```
```suggestion
    NCrystal::CachePtr dummy_cache;
    auto nc_energy = NCrystal::NeutronEnergy{p.E()};
    ncrystal_xs = ncrystal_mat_->crossSection(dummy_cache, nc_energy,  {p.u().x, p.u().y, p.u().z}).get();
```

paulromano commented on src/material.cpp:
```diff
@@ -830,10 +849,25 @@ void Material::calculate_neutron_xs(Particle& p) const
 
     // Calculate microscopic cross section for this nuclide
-    const auto& micro {p.neutron_xs(i_nuclide)};
+#ifndef NCRYSTAL
+    const
+#endif
+    auto& micro {p.neutron_xs(i_nuclide)};
```
I would prefer the more-verbose and more-obvious:
```suggestion
#ifdef NCRYSTAL
    auto& micro {p.neutron_xs(i_nuclide)};
#else
    const auto& micro {p.neutron_xs(i_nuclide)};
#endif
```

paulromano commented on src/physics.cpp:
```diff
@@ -140,8 +140,26 @@ void sample_neutron_reaction(Particle& p)
-  scatter(p, i_nuclide);
 
+#ifdef NCRYSTAL
+  if (model::materials[p.material()]->m_NCrystal_mat() != nullptr && p.E() < settings::ncrystal_max_energy){
```
```suggestion
  if (model::materials[p.material()]->ncrystal_mat() && p.E() < settings::ncrystal_max_energy){
```

paulromano commented on src/material.cpp:
```diff
@@ -792,6 +800,17 @@ void Material::calculate_neutron_xs(Particle& p) const
+#ifdef NCRYSTAL
+  double ncrystal_xs = -1;
+
+  if (m_NCrystal_mat_ != nullptr && p.E() < settings::ncrystal_max_energy){
```
```suggestion
  if (ncrystal_mat_ && p.E() < settings::ncrystal_max_energy){
```

paulromano commented on src/physics.cpp:
```diff
@@ -140,8 +140,26 @@ void sample_neutron_reaction(Particle& p)
+  if (model::materials[p.material()]->m_NCrystal_mat() != nullptr && p.E() < settings::ncrystal_max_energy){
+    NcrystalRNG_Wrapper rng(p.current_seed()); // Initialize RNG
+    //create a cache pointer for multi thread physics
+    NCrystal::CachePtr dummyCache;//fixme: avoid recreating here (triggers malloc)
+    auto nc_energy = NCrystal::NeutronEnergy{p.E()};
+    auto outcome = model::materials[p.material()]->m_NCrystal_mat()->sampleScatter( dummyCache, rng , nc_energy, {p.u().x, p.u().y, p.u().z});
```
```suggestion
    NCrystal::CachePtr dummy_cache;
    auto nc_energy = NCrystal::NeutronEnergy{p.E()};
    auto outcome = model::materials[p.material()]->ncystal_mat()->sampleScatter(dummy_cache, rng , nc_energy, {p.u().x, p.u().y, p.u().z});
```

paulromano commented on src/physics.cpp:
```diff
@@ -140,8 +140,26 @@ void sample_neutron_reaction(Particle& p)
+    Direction u_old {p.u()};
+    p.u() = Direction(outcome.direction[0], outcome.direction[1], outcome.direction[2]);
+    p.mu() = u_old.dot(p.u());
+    p.event_mt() = ELASTIC;//fixme: define a new label for NCrystal?
```
This is fine for now and is consistent with what we do for normal S(a,b) scattering
```suggestion
    p.event_mt() = ELASTIC;
```

paulromano reviewed (changes requested):
@marquezj @tkittel Thanks once again for this PR. This change really isn't very intrusive at all and I think with a little restructuring, we might be able to make it even cleaner. Here is my first round of comments, mostly just about naming/style conventions. I haven't had a chance yet to actually try this out but plan on doing so next week.

@ameliajo I was wondering if you might be interested in taking a look at this PR too as our resident thermal scattering expert :smile:

tkittel commented:
So @paulromano, unless I overlooked something I believe @marquezj took care of all the changes requested so far (including applying clang-format).

paulromano commented on include/openmc/physics.h:
```diff
@@ -101,6 +105,26 @@ void sample_secondary_photons(Particle& p, int i_nuclide);
+//==============================================================================
+
+class NCrystalRNGWrapper : public NCrystal::RNGStream {
+  uint64_t* openmc_seed_;
```
This should probably be in the `private:` section?

tkittel replied:
Well, it technically is since the default access level of classes is private :-)

But we can certainly make an explicit `private:` section if you prefer.

paulromano replied:
Thanks! Yes, we generally follow the C++ core guidelines when possible:
[NL.16: Use a conventional class member declaration order](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rl-order)

tkittel replied:
So this was done now.

paulromano commented on CMakeLists.txt:
```diff
@@ -98,6 +99,14 @@ macro(find_package_write_status pkg)
+#===============================================================================
+
+if(OPENMC_USE_NCRYSTAL)
+  find_package(NCrystal REQUIRED)
```
```suggestion
  find_package(NCrystal REQUIRED)
  message(STATUS "Found NCrystal: ${NCrystal_DIR} (version ${NCrystal_VERSION})")
```

paulromano reviewed (changes requested):
Thanks @marquezj and @tkittel for the updates on this PR. In addition to the comments below, the other thing this PR needs in order to be considered for merging is tests/documentation. For testing, take a look at existing examples in the `tests/regression_tests` folder. At a minimum, we should have a test that utilizes this new capability and ensures that we get the expected answer. Along with that, we'll need some updates to the CI environment so that we can build against NCrystal -- you can look at `tools/ci/gha-install-njoy.sh` for an example of how we install NJOY. Adding something similar for NCrystal should be straightforward, and (correct me if I'm wrong) the build time for NCrystal shouldn't be prohibitive to add to CI.

I was going to ask for some example inputs using this so I can test it out, but I think the addition of a regression test will serve that purpose.

For documentation, the only thing that's absolutely required for now is to update the section of our user's guide mentioning optional dependencies.

tkittel commented:
FYI we haven't forgotten this, but some other work came in the way. We hope to spend some time on this on Friday.

tkittel commented:
I just pushed a few updates, mainly to use a new py API method I have just added in NCrystal 3.4.1, which helps getting the basic material composition in Python even for complicated (multiphase/enriched/...) materials. In the rare and somewhat academic case of materials where the same element now appears both as a natural element and as a explicit isotope, we are now using the openmc natural abundancies to break apart the natural element. This is like what we have been doing all along in Geant4 (with the C++ API), where we instead of course were asking Geant4 to provide the natural abundancies.

I also added `material_id` and `name` optional parameters on the `Material.from_ncrystal` method, which then simply pass these two parameters on to the Material constructor.

tkittel commented:
Outstanding issues are tests, documentation and CI as requested. We will have a look at those tomorrow.

tkittel commented:
@paulromano I added a `tools/ci/gha-install-ncrystal.sh` script now (can most likely be improved a bit in the future). Should we also update the `gha-install.sh` and `gha-install.py` files? Also, will it be possible to enable the workflows on the current PR, so we can try to get the new tests for NCrystal to work?

marquezj commented:
Hi, sorry for the delay. I added documentation and tests. @tkittel check the documentation: is brief but more can be found in the NCrystal repo and papers (I put links). I added one simple test to check it works. It will be easy to add tests with more complex config strings.

tkittel commented:
@marquezj it looks fine to me. Once we hopefully have some even better NCrystal documentation, we can go back to the openmc documentation and see if we need to update various links, or add new links. But with the current pile of items to do, it might not happen for quite some time :grin:

tkittel commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
+
+::
+   
+   mat = openmc.Material.from_ncrystal('Al_sg225.ncmat')
```
should we perhaps use `'Al_sg225.ncmat;temp=300K'` since many users most likely want to know how to set the temperature?

tkittel commented:
So the tests are unsurprisingly failing with:
```
 import NCrystal
 ModuleNotFoundError: No module named 'NCrystal'
```
@paulromano could you kindly suggest where you would like changes, to enable NCrystal being built? (btw. NCrystal is on conda-forge now, but I guess this CI setup is not conda-based?).

marquezj commented on cmake/OpenMCConfig.cmake.in:
```diff
@@ -9,6 +9,11 @@ if(@OPENMC_USE_DAGMC@)
 endif()
 
+if(@OPENMC_USE_NCRYSTAL@)
+  find_package(NCrystal REQUIRED)
```
Ok. Let's see what @paulromano thinks.

paulromano replied:
Interesting, I wasn't even aware of `find_dependency`. Let's keep it as `find_package` for now and perhaps in a future PR, we can change them all to `find_dependency`.

tkittel commented:
@paulromano : Is it normal that some of the checks do not start? All the started ones have now passed.
Are there any remaining things we need to look at, or are you happy to merge this now?

Cheers,
Thomas

paulromano commented:
@tkittel No, you don't need to worry about that. GitHub is expecting a set of test configurations that does not include the new ncrystal option, so that's why those extra ones are shown. Once this is merged, we'll just change the list of expected test configurations to include the ones with ncrystal included.

paulromano commented on CMakeLists.txt:
```diff
@@ -482,6 +492,13 @@ if (OPENMC_USE_MPI)
+  target_link_libraries(libopenmc NCrystal::NCrystal)
+endif()
+
+
+
```
```suggestion
```

paulromano commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
 
 .. _usersguide_naming:
 
+------------------
+Adding NCrystal materials
+------------------
```
```suggestion
-------------------------
Adding NCrystal materials
-------------------------
```

paulromano commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
+Additional support for thermal scattering can be added by using NCrystal_.
+The :meth:`Material.from_ncrystal` class method generates a :class:`openmc.Material` object from
+an `NCrystal configuration string <https://github.com/mctools/ncrystal/wiki/Using-NCrystal#uniform-material-configuration-syntax>`_.
+Temperature, material composition and density are passed from the configuration string
```
```suggestion
Temperature, material composition, and density are passed from the configuration string
```

paulromano commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
+an `NCrystal configuration string <https://github.com/mctools/ncrystal/wiki/Using-NCrystal#uniform-material-configuration-syntax>`_.
+Temperature, material composition and density are passed from the configuration string
+and the `NCMAT file <https://github.com/mctools/ncrystal/wiki/NCMAT-format>`_
+that define the material, e.g:
+
+::
+   
+   mat = openmc.Material.from_ncrystal('Al_sg225.ncmat;temp=300K')
```
```suggestion
that define the material, e.g.::

   mat = openmc.Material.from_ncrystal('Al_sg225.ncmat;temp=300K')
```

paulromano commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
+
+defines an oriented germanium single crystal with 40 arcsec mosaicity.
+
+NCrystal only handles low energy neutron interactions. Other interaction are
```
```suggestion
NCrystal only handles low energy neutron interactions. Other interactions are
```

paulromano commented on docs/source/usersguide/materials.rst:
```diff
@@ -101,6 +101,42 @@ you would need to add hydrogen and oxygen to a material and then assign the
+<https://github.com/mctools/ncrystal/wiki/Data-library>`_ but more materials can
+be added by creating NCMAT files or on-the-fly in the configuration string.
+
+.. warning:: Currently, NCrystal_ materials cannot be modified after created. 
```
```suggestion
.. warning:: Currently, NCrystal_ materials cannot be modified after they are created. 
```

paulromano commented on openmc/material.py:
```diff
@@ -331,6 +334,73 @@ def from_hdf5(cls, group: h5py.Group):
 
+    @classmethod
+    def from_ncrystal(cls, cfg, material_id=None, name=''):
+        """Create material from NCrystal configuration string. Density,
+        temperature, and material composition, and (ultimately) thermal neutron
+        scattering, will be automatically be provided by NCrystal based on this
+        string. The name and material_id parameters are simply passed on to the
+        Material constructor.
```
Docstrings should start with a one-line summary followed by extra details:
```suggestion
        """Create material from NCrystal configuration string.
        
        Density, temperature, and material composition, and (ultimately) thermal
        neutron scattering will be automatically be provided by NCrystal based
        on this string. The name and material_id parameters are simply passed on
        to the Material constructor.
```

paulromano commented on openmc/material.py:
```diff
@@ -331,6 +334,73 @@ def from_hdf5(cls, group: h5py.Group):
+        import NCrystal
+        nc_mat = NCrystal.createInfo(cfg)
+
+        def openmc_natabund( Z ):
+            #nc_mat.getFlattenedComposition might need natural abundancies.
+            #This call-back function is used so NCrystal can flatten composition
+            #using OpenMC's natural abundancies. In practice this function will
+            #only get invoked in the unlikely case where a material is specified
+            #by referring both to natural elements and specific isotopes of the
+            #same element.
+            elem_name = openmc.data.ATOMIC_SYMBOL.get( Z, None )
+            if not elem_name:
+                raise ValueError( f'Element with Z={Z} is not known' )
+            l = []
+            for iso_name,abund in openmc.data.isotopes( elem_name ):
+                l.append( ( int(iso_name[ len(elem_name) : ]), abund ) )
+            return l
```
```suggestion
        def openmc_natabund(Z):
            #nc_mat.getFlattenedComposition might need natural abundancies.
            #This call-back function is used so NCrystal can flatten composition
            #using OpenMC's natural abundancies. In practice this function will
            #only get invoked in the unlikely case where a material is specified
            #by referring both to natural elements and specific isotopes of the
            #same element.
            elem_name = openmc.data.ATOMIC_SYMBOL.get(Z)
            if not elem_name:
                raise ValueError( f'Element with Z={Z} is not known' )
            return [
                (int(iso_name[len(elem_name):]), abund)
                for iso_name, abund in openmc.data.isotopes(elem_name)
            ]
```

paulromano commented on openmc/material.py:
```diff
@@ -331,6 +334,73 @@ def from_hdf5(cls, group: h5py.Group):
+                l.append( ( int(iso_name[ len(elem_name) : ]), abund ) )
+            return l
+
+        flat_compos = nc_mat.getFlattenedComposition( preferNaturalElements = True,
+                                                      naturalAbundProvider = openmc_natabund )
+
+        # Create the Material
+        material = cls( material_id = material_id,
+                        name = name,
+                        temperature = nc_mat.getTemperature() )
+
+        for Z, A_vals in flat_compos:
+            elemname = openmc.data.ATOMIC_SYMBOL.get(Z,None)
+            if not elemname:
+                raise ValueError(f'Element with Z={Z} is not known')
+            for A, frac in A_vals:
+                if A:
+                    material.add_nuclide( elemname + str(A), frac, 'ao' )
+                else:
+                    material.add_element( elemname, frac, 'ao' )
+
+        material.set_density( 'g/cm3', nc_mat.getDensity() )
+        material._ncrystal_cfg = NCrystal.normaliseCfg( cfg )
```
Just noting here that Python code in OpenMC should follow the [PEP8 style guide](https://peps.python.org/pep-0008/)
```suggestion
        flat_compos = nc_mat.getFlattenedComposition(preferNaturalElements = True,
                                                      naturalAbundProvider=openmc_natabund)

        # Create the Material
        material = cls(material_id=material_id,
                       name=name,
                       temperature=nc_mat.getTemperature())

        for Z, A_vals in flat_compos:
            elemname = openmc.data.ATOMIC_SYMBOL.get(Z)
            if not elemname:
                raise ValueError(f'Element with Z={Z} is not known')
            for A, frac in A_vals:
                if A:
                    material.add_nuclide(f'{elemname}{A}', frac)
                else:
                    material.add_element(elemname, frac)

        material.set_density('g/cm3', nc_mat.getDensity())
        material._ncrystal_cfg = NCrystal.normaliseCfg(cfg)
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+        """Digest info in the statepoint and return as a string."""
+
+        # Read the statepoint file.
+        sp = openmc.StatePoint(self._sp_name)
+        tal = sp.get_tally(name='angular distribution')
+        df = tal.get_pandas_dataframe()
```
```suggestion
        with openmc.StatePoint(self._sp_name) as sp:
            tal = sp.get_tally(name='angular distribution')
            df = tal.get_pandas_dataframe()
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+
+    sample_sphere = openmc.Sphere(r=0.1)
+    outer_sphere = openmc.Sphere(r=100, boundary_type="vacuum")
+    cell1 = openmc.Cell(region= -sample_sphere, fill=m1)
+    cell2_region= +sample_sphere&-outer_sphere
+    cell2 = openmc.Cell(region= cell2_region, fill=None)
```
```suggestion
    cell1 = openmc.Cell(region=-sample_sphere, fill=m1)
    cell2_region= +sample_sphere & -outer_sphere
    cell2 = openmc.Cell(region=cell2_region, fill=None)
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+    cell1 = openmc.Cell(region= -sample_sphere, fill=m1)
+    cell2_region= +sample_sphere&-outer_sphere
+    cell2 = openmc.Cell(region= cell2_region, fill=None)
+    uni1 = openmc.Universe(cells=[cell1, cell2])
+    geometry = openmc.Geometry(uni1)
```
```suggestion
    geometry = openmc.Geometry([cell1, cell2])
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+    # Source definition
+
+    source = openmc.Source()
+    source.space = openmc.stats.Point(xyz = (0,0,-20))
+    source.angle = openmc.stats.Monodirectional(reference_uvw = (0,0,1))
```
```suggestion
    source.space = openmc.stats.Point((0, 0, -20))
    source.angle = openmc.stats.Monodirectional(reference_uvw=(0, 0, 1))
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+
+    tally1 = openmc.Tally(name="angular distribution")
+    tally1.scores = ["current"]
+    filter1 = openmc.filter.SurfaceFilter(sample_sphere)
+    filter2 = openmc.filter.PolarFilter(np.linspace(0,np.pi,180+1))
+    filter3 = openmc.filter.CellFromFilter(cell1)
```
```suggestion
    filter1 = openmc.SurfaceFilter(sample_sphere)
    filter2 = openmc.PolarFilter(np.linspace(0, np.pi, 180+1))
    filter3 = openmc.CellFromFilter(cell1)
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+    tally1.filters = [filter1, filter2, filter3]
+    tallies = openmc.Tallies([tally1])
+
+    return openmc.model.Model(geometry, materials, settings, tallies)
```
```suggestion
    return openmc.Model(geometry, materials, settings, tallies)
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+    reason="NCrystal materials are not enabled.")
+
+def compute_angular_distribution(cfg, E0, N):
+    """Return a openmc.model.Model() object for a monoenergetic pencil
```
```suggestion
    """Return an openmc.Model() object for a monoenergetic pencil
```

paulromano commented on tests/regression_tests/ncrystal/test.py:
```diff
@@ -0,0 +1,77 @@
+    not openmc.lib._ncrystal_enabled(),
+    reason="NCrystal materials are not enabled.")
+
+def compute_angular_distribution(cfg, E0, N):
```
A name like `pencil_beam_model` might better convey the fact that this returns a Model object

tkittel replied:
This was done now.

paulromano commented on tools/ci/gha-install.py:
```diff
@@ -54,6 +54,11 @@ def install(omp=False, mpi=False, phdf5=False, dagmc=False, libmesh=False):
 
+    if ncrystal:
+        cmake_cmd.append('-DOPENMC_USE_NCRYSTAL=ON')
+        ncrystal_cmake_path = os.environ.get('HOME')+'/ncrystal_inst/lib/cmake'
+        cmake_cmd.append('-DCMAKE_PREFIX_PATH=' + ncrystal_cmake_path)
```
```suggestion
        ncrystal_cmake_path = os.environ.get('HOME') + '/ncrystal_inst/lib/cmake'
        cmake_cmd.append(f'-DCMAKE_PREFIX_PATH={ncrystal_cmake_path}')
```

paulromano commented on tools/ci/gha-script.sh:
```diff
@@ -14,5 +14,12 @@ if [[ $EVENT == 'y' ]]; then
 
+# Check NCrystal installation
+if [[ $NCRYSTAL = 'y' ]]; then
+    # Change environmental variables
+    eval $( "${HOME}/ncrystal_inst/bin/ncrystal-config" --setup )
+    nctool --test
```
```suggestion
  # Change environmental variables
  eval $( "${HOME}/ncrystal_inst/bin/ncrystal-config" --setup )
  nctool --test
```

paulromano reviewed (changes requested):
Thanks @tkittel and @marquezj for the updates! It's looking much better than where it started. Here's another round of comments. I think we're getting pretty close at this point.

tkittel commented:
Thanks @paulromano , all changes committed as requested.

Anything else? Let me know if you want me to do a rebase of the branch on main now, or do any squashing or whatever :-)

tkittel commented:
Hi @paulromano 

Despite the status saying "1 change requested", I think we did everything. Let us know if you think more is needed :smile:

Cheers,
Thomas

paulromano commented:
@tkittel Apologies this fell off my radar -- will take another look soon!

paulromano commented:
@tkittel One final question -- should `ncrystal_max_energy` be configurable? Right now it defaults to 5 eV and it doesn't appear that there's any way to change this.

tkittel commented:
I would postpone that for another time, since there is a common issue here with e.g. the Geant4 hooks. Once we eventually look into it, it might anyway be that we go for another solution altogether for this threshold (e.g. a per-material threshold rather than a global one).

paulromano reviewed (approved):
Excited to have this merged in -- thanks @tkittel @marquezj for this great contribution!

tkittel commented:
Very exciting to us as well, thanks a lot @paulromano for the support and shepherding! :smile:
